test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Build.AspNet/BulkOperation/BulkOperationController.cs

```diff
@@ -0,0 +1,259 @@
+            friends.Add(friend2);
+
+            emp1.TrySetPropertyValue("UnTypedFriends", friends);
+
+            var emp2 = new EdmEntityObject(entityType);
+            emp2.TrySetPropertyValue("ID", 2);
+            emp2.TrySetPropertyValue("Name", "Test2");
+
+            var friends2 = new List<EdmStructuredObject>();
+            var friend3 = new EdmEntityObject(friendType);
+            friend3.TrySetPropertyValue("Id", 3);
+            friend3.TrySetPropertyValue("Name", "Test3");
+
+            var friend4 = new EdmEntityObject(friendType);
+            friend4.TrySetPropertyValue("Id", 4);
+            friend4.TrySetPropertyValue("Name", "Test4");
+
+            friends2.Add(friend3);
+            friends2.Add(friend4);
+
+            emp2.TrySetPropertyValue("UnTypedFriends", friends2);
+
+            var emp3 = new EdmEntityObject(entityType);
+            emp3.TrySetPropertyValue("ID", 3);
+            emp3.TrySetPropertyValue("Name", "Test3");
+
+            var friends35 = new List<EdmStructuredObject>();
+            var friend5 = new EdmEntityObject(friendType);
+            friend5.TrySetPropertyValue("Id", 5);
+            friend5.TrySetPropertyValue("Name", "Test5");
+
+            friends35.Add(friend5);
+
+            emp3.TrySetPropertyValue("UnTypedFriends", friends35);
+
+            EmployeesTypeless.Add(emp1);
+            EmployeesTypeless.Add(emp2);
+            EmployeesTypeless.Add(emp3);
+        }
+
+        [EnableQuery(PageSize = 10, MaxExpansionDepth = 5)]
+        public ITestActionResult Get()
+        {
+            return Ok(Employees.AsQueryable());
+        }
+
+        [EnableQuery]
+        public ITestActionResult Get(int key)
+        {
+            var emp = Employees.SingleOrDefault(e => e.ID == key);
+            return Ok(emp);
+        }
+
+        [ODataRoute("Employees({key})/Friends")]
+        public ITestActionResult GetFriends(int key)
+        {
+            var emp = Employees.SingleOrDefault(e => e.ID == key);
+            return Ok(emp.Friends);
+        }
+
+        [ODataRoute("Employees({key})/UnTypedFriends")]
+        public ITestActionResult GetUnTypedFriends(int key)
+        {
+            var entity = Request.GetModel().FindDeclaredType("Microsoft.Test.E2E.AspNet.OData.BulkOperation.UnTypedEmployee") as IEdmEntityType;
+            InitTypeLessEmployees(entity);
+
+            foreach (var emp in EmployeesTypeless)
+            {
+                object obj;
+                emp.TryGetPropertyValue("ID", out obj);
+
+                if (Equals(key, obj))
+                {
+                    object friends;
+                    emp.TryGetPropertyValue("UntypedFriends", out friends);
+                    return Ok(friends);
+                }
+            }
+            return Ok();
+        }
+
+        [ODataRoute("Employees")]
+        [HttpPost]
+        public ITestActionResult Post([FromBody] Employee employee)
+        {
+            InitEmployees();
```

Comment:
Is this the controller method that handles bulk post requests? How does the customer tell whether it's an insert or update? How do you access the navigation properties that were in the payload from here.

Reply:
A deep insert does not have update. It is just an insert and linking to existing entities. if the existing entity does not exist it is created then linked. We have the Api handlers PR where we are providing logic that will handle all that for the customer as long as they provide with the objects through their handlers. At the moment, with these changes, the customer will be doing all this for themselves. But if they added ODataIdContainer on their model then it will be populated and they can use that to tell.